drivers/audio/audio_dma.c

```diff
@@ -0,0 +1,607 @@
+        /* Provide our overall capabilities.  The interfacing software
+         * must then call us back for specific info for each capability.
+         */
+
+        caps->ac_channels = 2;       /* Stereo output */
+
+        switch (caps->ac_subtype)
+          {
+            case AUDIO_TYPE_QUERY:
+
+              /* We don't decode any formats!  Only something above us in
+               * the audio stream can perform decoding on our behalf.
+               */
+
+              /* The types of audio units we implement */
+
+              if (audio_dma->playback)
+                caps->ac_controls.b[0] = AUDIO_TYPE_OUTPUT;
+              else
+                caps->ac_controls.b[0] = AUDIO_TYPE_INPUT;
+              caps->ac_format.hw = 1 << (AUDIO_FMT_PCM - 1);
+              break;
+
+            default:
+              caps->ac_controls.b[0] = AUDIO_SUBFMT_END;
+              break;
+          }
+
+        break;
+
+        /* Provide capabilities of our OUTPUT unit */
+
+      case AUDIO_TYPE_OUTPUT:
+      case AUDIO_TYPE_INPUT:
+
+        caps->ac_channels = 2;
+
+        switch (caps->ac_subtype)
+          {
+            case AUDIO_TYPE_QUERY:
+
+            /* Report the Sample rates we support */
+
+              caps->ac_controls.hw[0] = AUDIO_SAMP_RATE_8K  |
+                                        AUDIO_SAMP_RATE_11K |
+                                        AUDIO_SAMP_RATE_16K |
+                                        AUDIO_SAMP_RATE_22K |
+                                        AUDIO_SAMP_RATE_32K |
+                                        AUDIO_SAMP_RATE_44K |
+                                        AUDIO_SAMP_RATE_48K |
+                                        AUDIO_SAMP_RATE_96K |
+                                        AUDIO_SAMP_RATE_128K |
+                                        AUDIO_SAMP_RATE_160K |
+                                        AUDIO_SAMP_RATE_172K |
+                                        AUDIO_SAMP_RATE_192K;
+              break;
+          }
+
+        break;
+   }
+
+  /* Return the length of the audio_caps_s struct for validation of
+   * proper Audio device type.
+   */
+
+  return caps->ac_len;
+}
+
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+static int audio_dma_configure(struct audio_lowerhalf_s *dev,
+                               void *session,
+                               const struct audio_caps_s *caps)
+#else
+static int audio_dma_configure(struct audio_lowerhalf_s *dev,
+                               const struct audio_caps_s *caps)
+#endif
+{
+  struct audio_dma_s *audio_dma = (struct audio_dma_s *)dev;
+  struct dma_config_s cfg;
+  int ret = -EINVAL;
+
+  DEBUGASSERT(audio_dma && caps);
+  audinfo("ac_type: %d\n", caps->ac_type);
+
+  /* Process the configure operation */
+
+  switch (caps->ac_type)
+    {
+      case AUDIO_TYPE_OUTPUT:
+        if (audio_dma->playback)
+          {
+            memset(&cfg, 0, sizeof(struct dma_config_s));
+            cfg.direction = DMA_MEM_TO_DEV;
```

Comment:
Where `DMA_MEM_TO_DEV` is defined?

Reply:
It's defined in https://github.com/apache/nuttx/pull/8773.